charger/smaevcharger.go

```diff
@@ -0,0 +1,374 @@
+}
+
+func init() {
+	registry.Add("smaevcharger", NewSmaevchargerFromConfig)
+}
+
+// NewSmaevchargerFromConfig creates a Smaevcharger charger from generic config
+func NewSmaevchargerFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		Host     string
+		User     string
+		Password string
+	}{}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.Host == "" {
+		return nil, errors.New("missing host")
+	}
+	if cc.User == "" {
+		return nil, errors.New("missing user")
+	} else if cc.User == "admin" {
+		return nil, errors.New("user admin not allowed, create new user")
+	}
+	if cc.Password == "" {
+		return nil, errors.New("missing password")
+	}
+
+	return NewSmaevcharger(cc.Host, cc.User, cc.Password)
+}
+
+// NewSmaevcharger creates Smaevcharger charger
+func NewSmaevcharger(host string, user string, password string) (api.Charger, error) {
+	wb := &Smaevcharger{
+		Helper:		request.NewHelper(util.NewLogger("smaevcharger")),
+		host:      	"http://" + host + "/api/v1",
+		user:     	user,
+		password: 	password,
+	}
+
+	return wb, nil
+}
+
+// Status implements the api.Charger interface
+func (wb *Smaevcharger) Status() (api.ChargeStatus, error) {
+	StateChargerCharging := wb.GetMeasurement("Measurement.Operation.EVeh.ChaStt")
+
+	switch StateChargerCharging {
+	case smaevchNConNCarNChar:// No Car connectec and no charging
+		return api.StatusA, nil
+	case smaevchYConYCarNChar:// Car connected and no charging
+		return api.StatusB, nil
+	case smaevchYConYCarYChar:// Car connected and charging
+		return api.StatusC, nil
+	}
+		return api.StatusNone, fmt.Errorf("unkown charger state: %s", StateChargerCharging)
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Smaevcharger) Enabled() (bool, error) {
+	StateChargerMode := wb.GetParameter("Parameter.Chrg.ActChaMod")
+
+	switch StateChargerMode {
+	case smaevchFastCharge:// Schnellladen - 4718
+		return true, nil
+	case smaevchOptiCharge:// Optimiertes Laden - 4719
+		return true, nil
+	case smaevchPlanCharge:// Laden mit Vorgabe - 4720
+		return true, nil
+	case smaevchStopCharge:// Ladestopp - 4721
+		return false, nil
+	}
+	return false, fmt.Errorf("unknown charger charge mode: %s", StateChargerMode)
+}
+
+// Enable implements the api.Charger interface
+func (wb *Smaevcharger) Enable(enable bool) error {
+	StateChargerSwitch := wb.GetMeasurement("Measurement.Chrg.ModSw")
+	if enable {
+		switch StateChargerSwitch {
+		case smaevchSwitchOeko:// Switch PV Loading
+			wb.SendParameter("Parameter.Chrg.ActChaMod", smaevchOptiCharge)
+		case smaevchSwitchFast:// Fast charging
+			wb.SendParameter("Parameter.Chrg.ActChaMod", smaevchFastCharge)
+		}
+	} else {
+		wb.SendParameter("Parameter.Chrg.ActChaMod", smaevchStopCharge)
+	}
+	return nil
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Smaevcharger) MaxCurrent(current int64) error {
+
+	//Convert current to watts
+	ChargeRate := current * 3 * 230

```

Comment:
Does not make sense for 1P/2P Loading, any suggestions?

Reply:
at least 1P loading? as this is also used with SMA HM 2.0

Reply:
I was unable to find any parameters that would enable a remote control of the used Phases. This seems to be an automated feature or this might be restricted to the SMA speedwire protocol. Indication of the used Phases could be done via the measured Current on each Phase as a  workaround.

Reply:
Evcc has current+phases1p3p api. Both should- combined use- set the power and Sma should adjust the phases. Just an idea…

Reply:
Found another option to Limit the MaxCurrent with the "EV Charger Supply" Current. I'll need to test different setups and charger behaviors. I'll probably test that with the [charger_ramp](https://github.com/evcc-io/evcc/blob/master/cmd/charger_ramp.go) as soon as i find some time for that. I'll check the resulting PWM Signal from the Charger and will post the findings.

Reply:
add support for 10mA steps MaxCurrentMillis
changed parameters / behavior of current limiter
Tested Setup with EVCC ramp function and Renault:
![image](https://user-images.githubusercontent.com/6517959/165648125-8ce31245-b68b-495d-9e80-167912e43e18.png)